network/ipfs_impl.go

```diff
@@ -86,24 +88,128 @@ type impl struct {
 }
 
 type streamMessageSender struct {
-	s     network.Stream
-	bsnet *impl
+	to     peer.ID
+	stream network.Stream
+	bsnet  *impl
+	opts   *MessageSenderOpts
+	done   chan struct{}
 }
 
-func (s *streamMessageSender) Close() error {
-	return helpers.FullClose(s.s)
+// Open a stream to the remote peer
+func (s *streamMessageSender) Connect(ctx context.Context) (network.Stream, error) {
+	if s.stream != nil {
+		return s.stream, nil
+	}
+
+	if err := s.bsnet.ConnectTo(ctx, s.to); err != nil {
+		return nil, err
+	}
+
+	// Check if the sender has been closed
+	select {
+	case <-s.done:
```

Comment:
Won't the context handle this?

Reply:
This _plus_ a context seems like overkill. But I'm not sure what the best design is.

Reply:
I agree - it's a consequence of having two ways to shutdown in the message queue, I will remove that and simplify here